File: sdk/node/src/types/webhooks.ts
export interface CreateWebhookRequest {
  name: string;
  url: string;
  events: string[];
  is_enabled?: boolean;
  retry_policy?: Record<string, any>;
  custom_headers?: Record<string, string>;
}

export interface UpdateWebhookRequest {
  name?: string;
  url?: string;
  events?: string[];
  is_enabled?: boolean;
  retry_policy?: Record<string, any>;
  custom_headers?: Record<string, string>;
}

export interface WebhookResponse {
  id: string;
  organization_id: string;
  name: string;
  url: string;
  events: string[];
  is_enabled: boolean;
  retry_policy?: Record<string, any>;
  custom_headers?: Record<string, string>;
  total_deliveries?: number;
  successful_deliveries?: number;
  failed_deliveries?: number;
  last_triggered_at?: string;
  created_at?: string;
  updated_at?: string;
}

export interface WebhookWithSecretResponse extends WebhookResponse {
  secret: string;
}

export interface TestWebhookRequest {
  event_type?: string;
}

export interface TestWebhookResponse {
  success: boolean;
  status_code?: number;
  response_time_ms?: number;
  response_body?: string;
  error?: string;
}

export interface WebhookDelivery {
  id: string;
  webhook_id: string;
  event_id: string;
  event_type: string;
  payload?: Record<string, any>;
  status: string;
  response_status?: number;
  response_time_ms?: number;
  attempt_count?: number;
  max_attempts?: number;
  next_retry_at?: string;
  last_error?: string;
  delivered_at?: string;
  failed_at?: string;
  created_at?: string;
}

export interface DeliveryQuery {
  event_type?: string;
  status?: string;
  limit?: number;
  offset?: number;
}

export interface WebhookEventType {
  id: string;
  category: string;
  description: string;
  payload_schema?: Record<string, any>;
  created_at?: string;
}

File: sdk/python/coreauth/types/webhooks.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel


class CreateWebhookRequest(BaseModel):
    name: str
    url: str
    events: list[str] = []
    is_enabled: bool = False
    retry_policy: Optional[dict[str, Any]] = None
    custom_headers: Optional[dict[str, str]] = None


class UpdateWebhookRequest(BaseModel):
    name: Optional[str] = None
    url: Optional[str] = None
    events: Optional[list[str]] = None
    is_enabled: Optional[bool] = None
    retry_policy: Optional[dict[str, Any]] = None
    custom_headers: Optional[dict[str, str]] = None


class WebhookResponse(BaseModel):
    id: Optional[str] = None
    organization_id: Optional[str] = None
    name: Optional[str] = None
    url: Optional[str] = None
    events: Optional[list[str]] = None
    is_enabled: Optional[bool] = None
    retry_policy: Optional[dict[str, Any]] = None
    custom_headers: Optional[dict[str, str]] = None
    total_deliveries: Optional[int] = None
    successful_deliveries: Optional[int] = None
    failed_deliveries: Optional[int] = None
    last_triggered_at: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class WebhookWithSecretResponse(BaseModel):
    id: Optional[str] = None
    organization_id: Optional[str] = None
    name: Optional[str] = None
    url: Optional[str] = None
    events: Optional[list[str]] = None
    is_enabled: Optional[bool] = None
    retry_policy: Optional[dict[str, Any]] = None
    custom_headers: Optional[dict[str, str]] = None
    total_deliveries: Optional[int] = None
    successful_deliveries: Optional[int] = None
    failed_deliveries: Optional[int] = None
    last_triggered_at: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    secret: Optional[str] = None


class TestWebhookRequest(BaseModel):
    event_type: Optional[str] = None


class TestWebhookResponse(BaseModel):
    success: bool
    status_code: Optional[int] = None
    response_time_ms: Optional[int] = None
    response_body: Optional[str] = None
    error: Optional[str] = None


class WebhookDelivery(BaseModel):
    id: Optional[str] = None
    webhook_id: Optional[str] = None
    event_id: Optional[str] = None
    event_type: Optional[str] = None
    payload: Optional[dict[str, Any]] = None
    status: Optional[str] = None
    response_status: Optional[int] = None
    response_time_ms: Optional[int] = None
    attempt_count: Optional[int] = None
    max_attempts: Optional[int] = None
    next_retry_at: Optional[str] = None
    last_error: Optional[str] = None
    delivered_at: Optional[str] = None
    failed_at: Optional[str] = None
    created_at: Optional[str] = None


class DeliveryQuery(BaseModel):
    event_type: Optional[str] = None
    status: Optional[str] = None
    limit: Optional[int] = None
    offset: Optional[int] = None


class WebhookEventType(BaseModel):
    id: Optional[str] = None
    category: Optional[str] = None
    description: Optional[str] = None
    payload_schema: Optional[dict[str, Any]] = None
    created_at: Optional[str] = None
